forc-plugins/forc-doc/src/render/mod.rs

```diff
@@ -0,0 +1,279 @@
+    match doc_links.get_mut(&key) {
+        Some(links) => links.push(doc.link()),
+        None => {
+            doc_links.insert(key, vec![doc.link()]);
+        }
+    }
+}
+fn populate_decls(
+    doc: &Document,
+    module_map: &mut BTreeMap<ModulePrefixes, BTreeMap<BlockTitle, Vec<DocLink>>>,
+) {
+    let module_prefixes = &doc.module_info.module_prefixes;
+    match module_map.get_mut(module_prefixes) {
+        Some(doc_links) => populate_doc_links(doc, doc_links),
+        None => {
+            let mut doc_links = BTreeMap::<BlockTitle, Vec<DocLink>>::new();
+            populate_doc_links(doc, &mut doc_links);
+            module_map.insert(module_prefixes.clone(), doc_links);
+        }
+    }
+}
+fn populate_modules(
+    doc: &Document,
+    module_map: &mut BTreeMap<ModulePrefixes, BTreeMap<BlockTitle, Vec<DocLink>>>,
+) {
+    let mut module_clone = doc.module_info.clone();
+    while module_clone.parent().is_some() {
+        let html_filename = if module_clone.depth() > 2 {
+            format!("{}/{INDEX_FILENAME}", module_clone.location())
+        } else {
+            INDEX_FILENAME.to_string()
+        };
+        let module_link = DocLink {
+            name: module_clone.location().to_owned(),
+            module_info: module_clone.to_owned(),
+            html_filename,
+            preview_opt: doc.module_info.preview_opt(),
+        };
+        let module_prefixes = module_clone
+            .module_prefixes
+            .clone()
+            .split_last()
+            .unwrap()
+            .1
+            .to_vec();
+        match module_map.get_mut(&module_prefixes) {
+            Some(doc_links) => match doc_links.get_mut(&BlockTitle::Modules) {
+                Some(links) => {
+                    if !links.contains(&module_link) {
+                        links.push(module_link)
+                    }
+                }
+                None => {
+                    doc_links.insert(BlockTitle::Modules, vec![module_link]);
+                }
+            },
+            None => {
+                let mut doc_links: BTreeMap<BlockTitle, Vec<DocLink>> = BTreeMap::new();
+                doc_links.insert(BlockTitle::Modules, vec![module_link]);
+                module_map.insert(module_prefixes.clone(), doc_links);
+            }
+        }
+        module_clone.module_prefixes.pop();
+    }
+}
+fn populate_all_doc(doc: &Document, all_docs: &mut DocLinks) {
+    populate_doc_links(doc, &mut all_docs.links);
+}
+
+/// The finalized HTML file contents.
+#[derive(Debug)]
+pub(crate) struct HTMLString(pub(crate) String);
+impl HTMLString {
+    /// Final rendering of a [Document] HTML page to String.
+    fn from_rendered_content(rendered_content: Box<dyn RenderBox>) -> Self {
+        let markup = html! {
+            : doctype::HTML;
+            html {
+                : rendered_content
+            }
+        };
+
+        Self(markup.into_string().unwrap())
```

Comment:
Can we unwrap here or should we expect?

Reply:
I can probably do you one better and add an `anyhow!` error message